### PR TITLE
Fix GpuSoftmax and GpuSoftmaxWithBias for non-float32 operation.

### PR DESCRIPTION
**Pull request by abergeron**
This fixes a bug which causes crashes in float16 and wrong results in float64.

### CODE REVIEW

**Comment by nouiz** [comment]
Can you modify a current test to test float16 and float32?

**Comment by abergeron** [comment]
Done.

File: theano/sandbox/gpuarray/tests/test_nnet.py [hunk 1]
```diff
@@ -209,11 +207,7 @@ def softmax_with_bias_unittest_template(dtypeInput, dtypeBias):
...
     def cmp(n, m):
-        # print "test_softmax",n,m
-        if dtypeInput == 'float32':
-            data = numpy.arange(n * m, dtype='float32').reshape(n, m)
-        elif dtypeInput == 'float64':
-            data = numpy.arange(n * m, dtype='float64').reshape(n, m)
+        data = numpy.arange(n * m, dtype=dtypeInput).reshape(n, m)
 
         out = f(data)
         gout = f_gpu(data)
```
Comment (nouiz) [inline theano/sandbox/gpuarray/tests/test_nnet.py]: The assert bellow is failing as the comparison is too strict. Use:

theano.tests.unittest_tools.assert_allclose(out, gout)

Comment (abergeron) [inline theano/sandbox/gpuarray/tests/test_nnet.py]: Done.

Comment (nouiz) [inline theano/sandbox/gpuarray/tests/test_nnet.py]: test fails if run locally with wrong value and low error and nan.

**Comment by abergeron** [comment]
This should fix the nan problems with both tests.  I had to modify how they work a bit so make sure they still test what they should.

**Pull request merged**